File: src/integration/wirespec/src/jvmMain/java/community/flock/wirespec/java/serde/DefaultParamSerialization.java
package community.flock.wirespec.java.serde;

import community.flock.wirespec.java.Wirespec;

import java.lang.reflect.Type;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static community.flock.wirespec.java.serde.DefaultSerialization.*;

public interface DefaultParamSerialization extends Wirespec.ParamSerialization {

    @Override
    default <T> List<String> serializeParam(T value, Type type) {
        if (isList(type)) {
            return ((List<?>) value).stream()
                    .map(Object::toString)
                    .toList();
        }
        if (isOptional(type)) {
            return ((Optional<?>) value)
                    .map(it -> Collections.singletonList(it.toString()))
                    .orElseGet(Collections::emptyList);
        }
        return Collections.singletonList(value.toString());
    }

    @Override
    @SuppressWarnings("unchecked")
    default <T> T deserializeParam(List<String> values, Type type) {
        if (isList(type)) {
            return (T) deserializeList(values, type);
        }
        if (isOptional(type)) {
            return (T) deserializeOptional(values, type);
        }
        Class<?> rawType = getRawType(type);
        if (isWirespecEnum(rawType)) {
            return (T) deserializeEnum(values, rawType);
        }
        return (T) deserializePrimitive(values, rawType);
    }


}

File: src/integration/wirespec/src/jvmMain/kotlin/community/flock/wirespec/kotlin/serde/DefaultParamSerialization.kt
package community.flock.wirespec.kotlin.serde

import community.flock.wirespec.kotlin.Wirespec
import community.flock.wirespec.kotlin.serde.DefaultSerialization.deserializeEnum
import community.flock.wirespec.kotlin.serde.DefaultSerialization.deserializeList
import community.flock.wirespec.kotlin.serde.DefaultSerialization.deserializePrimitive
import community.flock.wirespec.kotlin.serde.DefaultSerialization.isList
import community.flock.wirespec.kotlin.serde.DefaultSerialization.isWirespecEnum
import kotlin.reflect.KType

public class DefaultParamSerialization : Wirespec.ParamSerialization {

    override fun <T : Any> serializeParam(value: T, type: KType): List<String> = when {
        type.isList() -> (value as List<*>).map { it.toString() }
        else -> listOf(value.toString())
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : Any> deserializeParam(values: List<String>, type: KType): T = when {
        type.isList() -> deserializeList(values, type)
        type.isWirespecEnum() -> deserializeEnum(values, type)
        else -> deserializePrimitive(values, type)
    } as T
}
